src/charmed_kubeflow_chisme/testing/cos_integration.py

```diff
@@ -0,0 +1,226 @@
+) -> None:
+    """Deploy grafana-agent-k8 and add relate it with app.
+
+    Helper function to deploy and relate grafana-agent-k8s with provided app.
+
+    Args:
+        model: libjuju model object
+        app: name of application with which the Grafana agent should be related
+        channel: grafana-agent-k8s channel name, latest/stable as default
+        dashboard: relate <app>:grafana-dashboard grafana-agent-k8s:grafana-dashboards-consumer
+        metrics: relate <app>:metrics-endpoint grafana-agent-k8s:metrics-endpoint
+    """
+    assert app in model.applications, f"application {app} was not found"
+
+    log.info("deploying %s from %s channel", GRAFANA_AGENT_APP, channel)
+    await model.deploy(GRAFANA_AGENT_APP, channel=channel)
+
+    if dashboard is True:
+        log.info(
+            "juju relate %s:%s %s:grafana-dashboards-consumer",
+            app,
+            GRAFANA_DASHBOARD_RELATION,
+            GRAFANA_AGENT_APP,
+        )
+        await model.add_relation(
+            f"{app}:{GRAFANA_DASHBOARD_RELATION}",
+            f"{GRAFANA_AGENT_APP}:grafana-dashboards-consumer",
+        )
+
+    if metrics is True:
+        log.info(
+            "juju relate %s:%s %s:metrics-endpoint",
+            app,
+            METRICS_ENDPOINT_RELATION,
+            GRAFANA_AGENT_APP,
+        )
+        await model.add_relation(
+            f"{app}:{METRICS_ENDPOINT_RELATION}", f"{GRAFANA_AGENT_APP}:metrics-endpoint"
+        )
+
+    await model.wait_for_idle(apps=[GRAFANA_AGENT_APP], status="blocked", timeout=5 * 60)
+
+
+async def _check_metrics_endpoint(app: Application, metrics_endpoint: str) -> None:
+    """Check metrics endpoint accessibility.
+
+    Checking accessibility of metrics endpoint from grafana-agent-k8s. If metrics endpoint is
+    defined as `*:5000/metrics` it will be changed to `<app-name>.<namespace>.svc:5000/metrics`.
+    """
+    if metrics_endpoint.startswith("*"):
+        url = f"http://{app.name}.{app.model.name}.svc{metrics_endpoint[1:]}"
+    else:
+        url = f"http://{metrics_endpoint}"
+
+    cmd = f"curl -m 5 -sS {url}"
+    grafana_agent_app = app.model.applications[GRAFANA_AGENT_APP]
+    log.info("testing metrics endpoint with cmd: `%s`", cmd)
+    for unit in grafana_agent_app.units:
+        await _run_on_unit(unit, cmd)
+
+
+async def _get_app_relation_data(app: Application, endpoint_name: str) -> Dict[str, Any]:
+    """Get relations from endpoint name."""
+    assert len(app.units) > 0, f"application {app.name} has no units"
+    unit = app.units[0]  # Note(rgildein) use first unit, since we are getting application data
+    relations = [
+        relation
+        for relation in app.relations
+        if any(endpoint.name == endpoint_name for endpoint in relation.endpoints)
+    ]
+    log.info("found relations %s for %s:%s", relations, app.name, endpoint_name)
+
+    assert len(relations) == 1, f"{endpoint_name} is missing or there are too many of them"
```

Comment:
making the distinction here can be helpful, whether the endpoint is missing or there is more than one
I realize we can conclude that from the log above, but just a suggestion to improve the debugging for us.